core/src/consensus.rs

```diff
@@ -520,9 +530,34 @@ impl Tower {
                             // all of them.
                             panic!("no ancestors found with slot: {}", last_voted_slot);
                         } else {
-                            // bank_forks doesn't have corresponding data for the stray restored last vote,
-                            // meaning some inconsistency between saved tower and ledger.
-                            // (newer snapshot, or only a saved tower is moved over to new setup?)
+                            // This condition shouldn't occur under normal validator operation, indicating
+                            // something unusual happened.
```

Comment:
For the record, this is the well-written proof of this:


> pleb:
> This is my line of reasoning/understanding of why that log shouldn't ever trigger under normal operation.
> 
>  https://github.com/solana-labs/solana/pull/12875/files#diff-0b654209e209924bfc70e8dd4c84d0979e36b6ca76e85a5899ba4572521a8170R550, so here: last_voted_slot is a stray with no ancestors.
> 
> So assume last_voted_slot is a stray with no ancestors. Assume that when the vote for last_voted_slot was made, the snapshot was S_old. So when we restart from a snapshot S:
> 
> 1) S we start from is > last_voted_slot. But then the warning added in that PR will not trigger because switch_slot > S > last_voted_slot
> 2) If the snapshot S <= last_voted_slot then 
>          a) if S >= S_old is the latest snapshot across all restarts, then you're guaranteed to have ancestors (unless something went wrong) because last_voted_slot is descended from S and the ledger should contain all the descendants of S. 
>          b) If S < S_old, the do_adjust_lockouts_after_replay() should panic because it can't find any anchored slots
> 
> So in all cases under normal operation, that log shouldn't tigger, is that right?


And this proof is correct as far as I peer-reviewed this.